// koa-ejs-remote/example/app.js
/*!
 * koa-ejs - example/app.js
 * Copyright(c) 2017 dead_horse <dead_horse@qq.com>
 * MIT Licensed
 */

'use strict';

/**
 * Module dependencies.
 */

const Koa = require('koa');
const render = require('..');
const path = require('path');

const app = new Koa();

app.use(require('koa-static')(__dirname + '/view'));


render(app, {
  root: '/',
  host: 'http://localhost:7001',
  layout: 'template',
  viewExt: 'html',
  cache: false,
  debug: true,
  compileDebug:true
  // async: true
});

app.use(function (ctx, next) {
  ctx.state = ctx.state || {};
  ctx.state.now = new Date();
  ctx.state.ip = ctx.ip;
  ctx.state.version = '2.0.0';
  return next();
});

app.use(async function (ctx, next) {
  if (ctx.url === '/') {
    const users = [{ name: 'Dead Horse' }, { name: 'Jack' }, { name: 'Tom' }];
    await ctx.render('content', {
      users
    });
  }
  await next();
});

if (process.env.NODE_ENV === 'test') {
  module.exports = app.callback();
} else {
  app.listen(7001);
  console.log('open http://localhost:7001');
}



app.on('error', function (err) {
  console.log(err.stack);
});
