// koa-ejs-remote/index.js
/*!
 * koa-ejs-remote - index.js
 * MIT Licensed
 */

'use strict';

/**
 * Module dependencies.
 */

const debug = require('debug')('koa-ejs');
const fs = require('mz').fs;
const path = require('path');
const ejs = require('./lib/ejs');
const axios = require('axios');
const httpRetry = require('./lib/httpRetry');
const http = require('http');
const ejsLint = require('ejs-lint');

/**
 * Temp assigned for override later
 */
const parentResolveInclude = ejs.resolveInclude;


let httpClient = axios.create();
httpRetry(httpClient, { retries: 3 });
let httpAgent = new http.Agent({ keepAlive: true });



/**
 * default render options
 * @type {Object}
 */
const defaultSettings = {
  cache: true,
  layout: 'layout',
  viewExt: 'html',
  locals: {},
  compileDebug: false,
  debug: false,
  writeResp: true
};


/**
 * set app.context.render
 *
 * usage:
 * ```
 * await ctx.render('user', {name: 'dead_horse'});
 * ```
 * @param {Application} app koa application instance
 * @param {Object} settings user settings
 */
exports = module.exports = function (app, settings) {
  if (app.context.render) {
    return;
  }

  if (!settings || !settings.root) {
    throw new Error('settings.root required');
  }

  settings.root = path.resolve(process.cwd(), settings.root);

  //  load file from remote host if setting.host exists
  if (settings.host) {
    //override the ejs fileLoader
    ejs.fileLoader = async function httpFileLoader(path, options) {
      path = options.host + path;
      // let config = {
      //   url: path, method: 'get', httpAgent: httpAgent
      // };

      let response = await httpClient({ url: path, method: 'get', httpAgent: httpAgent });
      return response.data;//response.body;
    }
  }else if(settings.ejsLint){
    // load file from local, override the ejs fileLoader add ejs-lint
    // 2019-03-20 fengfeng
    ejs.fileLoader = async function readFileAsync(path) {
      return new Promise((resolve, reject) => {
        fs.readFile(path, 'utf8', (error, result) => {
          const ejsError = ejsLint.lint(result);
          if (error || ejsError) {
            return reject(error || ejsError);
          }
          resolve(result);
        });
      });
    }
  }
  /**
   * cache the generate package
   * @type {Object}
   */
  const cache = Object.create(null);

  settings = Object.assign({}, defaultSettings, settings);

  settings.viewExt = settings.viewExt
    ? '.' + settings.viewExt.replace(/^\./, '')
    : '';

  /**
   * generate html with view name and options
   * @param {String} view
   * @param {Object} options
   * @return {String} html
   */
  async function render(view, options) {
    view += settings.viewExt;
    const viewPath = path.join(settings.root, view);
    debug(`render: ${viewPath}`);
    // get from cache
    if (settings.cache && cache[viewPath]) {
      return cache[viewPath].call(options.scope, options);
    }

    const tpl = await ejs.fileLoader(settings.host + viewPath, options)

    // override `ejs` node_module `resolveInclude` function
    ejs.resolveInclude = function (name, filename, isDir) {
      if (!path.extname(name)) {
        name += settings.viewExt;
      }
      return parentResolveInclude(name, filename, isDir);
    }

    const fn = await ejs.compile(tpl, {
      filename: viewPath,
      _with: settings._with,
      compileDebug: settings.debug && settings.compileDebug,
      debug: settings.debug,
      delimiter: settings.delimiter,
      cache: settings.cache,
      host: settings.host,
      async: true
    });
    if (settings.cache) {
      cache[viewPath] = fn;
    }
    return fn.call(options.scope, options);
  }


  app.context.render = async function (view, _context) {
    const ctx = this;

    const context = Object.assign({}, ctx.state, _context);

    let html = await render(view, context);

    const layout = context.layout === false ? false : (context.layout || settings.layout);
    if (layout) {
      // if using layout
      context.body = html;
      html = await render(layout, context);
    }

    const writeResp = context.writeResp === false ? false : (context.writeResp || settings.writeResp);
    if (writeResp) {
      // normal operation
      ctx.type = 'html';
      ctx.body = html;
    } else {
      // only return the html
      return html;
    }
  };
};

/**
 * Expose ejs
 */

exports.ejs = ejs;
