// @things-factory/shell/dist-server/graphql/types/index.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const Domain = __importStar(require("./domain"));
const System = __importStar(require("./system"));
var filter_1 = require("./filter");
exports.Filter = filter_1.Filter;
var list_param_1 = require("./list-param");
exports.ListParam = list_param_1.ListParam;
var pagination_1 = require("./pagination");
exports.Pagination = pagination_1.Pagination;
var sorting_1 = require("./sorting");
exports.Sorting = sorting_1.Sorting;
var object_ref_1 = require("./object-ref");
exports.ObjectRef = object_ref_1.ObjectRef;
exports.queries = [Domain.Query];
exports.mutations = [Domain.Mutation];
exports.subscriptions = [System.Subscription];
exports.types = [...Domain.Types, ...System.Types];
